aptos-move/move-examples/rewards_pool/sources/rewards_pool.move

```diff
@@ -0,0 +1,211 @@
+/// An example module that manages rewards for multiple tokens on a per-epoch basis. Rewards can be added for multiple
+/// tokens by anyone for any epoch but only friended modules can increase/decrease claimer shares.
+///
+/// This module is designed to be integrated into a complete system that manages epochs and rewards.
+///
+/// The flow works as below:
+/// 1. A rewards pool is created with a set of reward tokens (fungible assets). If coins are to be used as rewards,
+/// developers can use the coin_wrapper module from move-examples/swap to convert coins into fungible assets.
+/// 2. Anyone can add rewards to the pool for any epoch for multiple tokens by calling add_rewards.
+/// 3. Friended modules can increase/decrease claimer shares for current epoch by calling increase_allocation and
+/// decrease_allocation.
+/// 4. Claimers can claim their rewards for any epoch that has ended by calling claim_rewards.
+///
+/// Note that due to integer division, there can be rounding errors that leave 1 unit of reward left for each token in
+/// an epoch's rewards after all claimers have claimed for that epoch. If this is undesirable, developers can add a skim
+/// function to allow extracting the "stuck" 1 unit of reward.
+module rewards_pool::rewards_pool {
+    use aptos_framework::fungible_asset::{Self, FungibleAsset, FungibleStore, Metadata};
+    use aptos_framework::object::{Self, Object, ExtendRef};
+    use aptos_std::pool_u64_unbound::{Self as pool_u64, Pool};
+    use aptos_std::smart_table::{Self, SmartTable};
+
+    use rewards_pool::epoch;
+
+    use std::signer;
+    use std::vector;
+
+    /// Rewards can only be claimed for epochs that have ended.
+    const EREWARDS_CANNOT_BE_CLAIMED_FOR_CURRENT_EPOCH: u64 = 1;
+    /// The rewards pool does not support the given reward token type.
+    const EREWARD_TOKEN_NOT_SUPPORTED: u64 = 2;
+    /// The caller doesn't have any pending rewards.
+    const ENO_REWARDS_TO_CLAIM: u64 = 3;
+
+    struct EpochRewards has store {
+        /// Total rewards for all reward tokens supported to be distributed for each epoch.
+        amount: SmartTable<u64, u64>,
+        /// As fungible assets are stored in separate store objects, we need to keep track of them.
+        store: Object<FungibleStore>,
+        /// We need to keep the fungible store's extend ref to be able to transfer rewards from it during claiming.
+        store_extend_ref: ExtendRef,
+    }
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct RewardsPool has key {
+        /// Track allocation per address per epoch. This is used to calculate the rewards for each address.
+        epoch_allocations: SmartTable<u64, Pool>,
+        epoch_rewards: SmartTable<Object<Metadata>, EpochRewards>,
+        reward_tokens: vector<Object<Metadata>>,
```

Comment:
why not smart?

Reply:
Not iterable. And this is not meant to support thousands of rewards tokens

Reply:
I think this should be matched with 48. smart+smart or simple_map+vector

Reply:
Why? If the data is not expected to be large, why does this need to use a smart vector?

Reply:
I mean,..., it's not large then why there is a smart table... vector + smart_table seems a little inconsistent.